mantidproject/mantid
# Add new integration algorithm for single-crystal Bragg peaks

### Description of work
Add new integration algorithm for single-crystal Bragg peaks adapted and extended from an algorithm in SXD2001 [1].
It uses scipy.minimize to fit peaks in adjacent pixels until the I/sigma ratio drops below some threshold - note scipy.minimize is used because mantid fitting currently does not perform well for this type of function - in principle it should be straightforward to use the mantid fitting at a later date.

[1] Gutmann, M. J. (2005). SXD2001. ISIS Facility, Rutherford Appleton Laboratory, Oxfordshire, England.

Fixes #33629

**Report to:** Matthias Gutmann (SXD)

### To test:

(1) Run this script (maybe delete a few rows/peaks from the table if it takes too long)
```
Load(Filename='/archive/ndxsxd/Instrument/data/cycle_10_3/SXD23767.raw', OutputWorkspace='SXD23767')
FindSXPeaksConvolve(InputWorkspace='SXD23767', PeaksWorkspace='peaks_out', GetNBinsFromBackToBackParams=True, ThresholdIoverSigma=20)
IntegratePeaks1DProfile(InputWorkspace='SXD23767', PeaksWorkspace='peaks_out', OutputWorkspace='peaks_int_poisson', GetNBinsFromBackToBackParams=True, NFWHM=6, CostFunction='Poisson', FixPeakParameters='A', FractionalChangeDSpacing=0.02, OutputFile='/mnt/babylon/Public/RWaite/SXD23767_1D_profile_poisson.pdf')
IntegratePeaks1DProfile(InputWorkspace='SXD23767', PeaksWorkspace='peaks_out', OutputWorkspace='peaks_int', GetNBinsFromBackToBackParams=True, NFWHM=6, CostFunction='RSq', FixPeakParameters='A', FractionalChangeDSpacing=0.02, OutputFile='/mnt/babylon/Public/RWaite/SXD23767_1D_profile_Rsq.pdf')
```
(2) Check the output pdf (it should have a couple of plots per peak)
![image](https://github.com/mantidproject/mantid/assets/55979119/cfa93b0c-37d5-41db-b9d6-c0362557623e)






---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
